### PR TITLE
MFA Sweep module + fixes 

### PR DESCRIPTION
**Pull request by ZephrFish**
## Additions
- Added in new user agents to match MFASweep, each will rotate through default recommended UAs from MS
- Added in MFASweep module + screenshots and docs to match, added line to config file too

## Updates
Added changes to readme to mirror the MFAModule
Updated Docker container to support ADAL/MSAL modules with Python 3.10
Updated requirements.txt with new modules required for MFASweep module

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,6 +76,26 @@ or
 python3 credmaster.py --config config.json
 ```
 
+### MFASweep Module/Flag
```
Comment (knavesec) [inline README.md:79]: Remove this section, readme is meant for a quick usage rather than a larger walkthrough. This would be better kept in the Wiki

File: README.md [hunk 2]
```diff
@@ -88,10 +108,11 @@ PRs welcome :)
 * New Plugin: [Optiv's Go365 Method](https://github.com/optiv/Go365) - Includes Office365 auth and userenum capabilities via SOAP
 * "Resume" functionality for paused/cancelled scans. Ideally storing data for APIs used, if they were destroyed and what user/pwd the spray was on
 * Method to reliably determine if an auth attempt was throttled, so the username could be re-queued and tried again later for full cover (would have to be per-plugin, return "throttled" boolean value in plugin script, requeue if throttled)
-* Notification system for webhooks (Teams TODO)
+* Notification system for webhooks (Always adding more)
 * Stop on success flag
 * Spray profile overhaul
 * Development notes
+* Limit MFA Sweep module to MSGraph/AZVault modules (Still Todo)
```
Comment (knavesec) [inline README.md:115]: Curious what you mean by this? Like only allow MFASweep if also running the MSGraph/AZVault modules? Would probably want the same ability to msol/o365? But IMO wherever creds are sprayed I think you should still be able to test them against MFASweep. Ex: If you spray an AD tied on-prem OWA server, get creds, then you want to see where that acct has access

File: passwords.txt [hunk 3]
```diff
@@ -1 +1 @@
-TestTest123
+TestTest123
```
Comment (knavesec) [inline passwords.txt:1]: How dareth thee

File: useragents.txt [hunk 4]
```diff
@@ -1,16 +1,11 @@
...
-Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0
-Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:59.0) Gecko/20100101 Firefox/59.0
-Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36
-Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36 Edge/16.16299
-Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/64.0.3282.186 Safari/537.36
-Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_4) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.1 Safari/605.1.15
-Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36
-Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:58.0) Gecko/20100101 Firefox/58.0
-Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36
+Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/107.0.0.0 Safari/537.36 Edg/107.0.1418.56
```
Comment (knavesec) [inline useragents.txt:1]: Why are these changed? Some are repeated, seems redundant

File: utils/mfasweep.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
+from datetime import datetime
```
Comment (knavesec) [inline utils/mfasweep.py:1]: Some overall thoughts: 

1. Biggest concern: Each of these MFA sweep requests will not be through a FireProx API, thus leaking your IP address. Generally fine, but poor opsec and not great for users who don't read the source
2. I sort of want to re-architect this in the future, ideally to have a "postex" type flag, almost like the plugins used. Would hopefully work in two ways, 1 used in-line during a spray so that a given postex module is ran on each set of creds found (like you have it set up), or 2 provide a username/password in the userpass file and _only_ run the postex module. Definitely a future plan, but just an FYI of my intentions

File: utils/mfasweep.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+def acquire_token_by_username_password(app, username, password):
+    """acquire_token_by_username_password() - See constraints here: https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-authentication-flows#constraints-for-ropc"""
+    pprint.pprint(
+        app.acquire_token_by_username_password(
+            username, password, scopes=_input_scopes(), app_name="My Signins"
+        )
+    )
+
+
+_JWK1 = """{"kty":"RSA", "n":"2tNr73xwcj6lH7bqRZrFzgSLj7OeLfbn8216uOMDHuaZ6TEUBDN8Uz0ve8jAlKsP9CQFCSVoSNovdE-fs7c15MxEGHjDcNKLWonznximj8pDGZQjVdfK-7mG6P6z-lgVcLuYu5JcWU_PeEqIKg5llOaz-qeQ4LEDS4T1D2qWRGpAra4rJX1-kmrWmX_XIamq30C9EIO0gGuT4rc2hJBWQ-4-FnE1NXmy125wfT3NdotAJGq5lMIfhjfglDbJCwhc8Oe17ORjO3FsB5CLuBRpYmP7Nzn66lRY3Fe11Xz8AEBl3anKFSJcTvlMnFtu3EpD-eiaHfTgRBU7CztGQqVbiQ", "e":"AQAB"}"""
```
Comment (knavesec) [inline utils/mfasweep.py:34]: Are these ever used? I don't see them anywhere in the code, curious what they are

File: utils/mfasweep.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+SSH_CERT_DATA = {"token_type": "ssh-cert", "key_id": "key1", "req_cnf": _JWK1}
+SSH_CERT_SCOPE = ["https://pas.windows.net/CheckMyAccess/Linux/.default"]
+
+
+# Check MFA credentials against various endpoints
+def mfasweep(username, password, notify_obj):
+    endpoint_table = [
```
Comment (knavesec) [inline utils/mfasweep.py:41]: Why is this whole table needed? The first number is just n+1 and the second appears to be a title, either way neither are used in the code. I think a simple list of URLs would work fine?

File: credmaster.py [hunk 8]
```diff
@@ -4,707 +4,965 @@
 from utils.fire import FireProx
 import utils.utils as utils
 import utils.notify as notify
+import utils.mfasweep as mfasweeper
```
Comment (knavesec) [inline credmaster.py:7]: There are only three places where code needs to be updated (input args, arg parsing, run mfasweep), I appreciate you taking the time to do formatting but it isn't necessary

File: README.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-# CredMaster #
+# CredMaster 2.0 #
```
Comment (knavesec) [inline README.md:1]: I dig haha

**Review by knavesec — CHANGES_REQUESTED** [review]
Love everything man, but there's a few places to be cleaned up

File: README.md (hunk 2 above)
Comment (ZephrFish) [reply README.md:115]: I meant in that if it's a MS module it'll give the option for MFASweep available rather than Okta etc :)

File: passwords.txt (hunk 3 above)
Comment (ZephrFish) [reply passwords.txt:1]: 😱

File: useragents.txt (hunk 4 above)
Comment (ZephrFish) [reply useragents.txt:1]: I changed the user agents as noticed that they were getting fingerprinted by some clients, so using more modern UAs to circumvent :)

File: utils/mfasweep.py (hunk 5 above)
Comment (ZephrFish) [reply utils/mfasweep.py:1]: 1. The implementation I looked into wasn't possible to flip through a fireprox endpoint due to the nature of how MFASweep is working on the backend :( thus this option, suppose could implement an opsec warning for users.
2. I agree a postex module or flag setup makes sense

File: utils/mfasweep.py (hunk 6 above)
Comment (ZephrFish) [reply utils/mfasweep.py:34]: Yes, when the function gets a valid set of creds that lack MFA they're used when a token is requested in the ADAL lib, I couldn't get it to work without them

File: utils/mfasweep.py (hunk 7 above)
Comment (ZephrFish) [reply utils/mfasweep.py:41]: Yeah a simple list would fix, I'll fix up and resubmit pull or can fix in dev whatever works easier

File: credmaster.py (hunk 8 above)
Comment (ZephrFish) [reply credmaster.py:7]: just wanted it to be tidy :)

File: README.md (hunk 9 above)
Comment (ZephrFish) [reply README.md:1]: Minor increment but it works :D

File: README.md (hunk 1 above)
Comment (ZephrFish) [reply README.md:79]: yeah I can move to wiki, I've not made changes to the wiki as been keeping as a readme locally instead so pushed up with the changes to add documentation.

Comment (knavesec) [reply README.md:79]: All good, no need to update the wiki asap since it's still Dev branch

File: README.md (hunk 2 above)
Comment (knavesec) [reply README.md:115]: IMO it should be available for anything, Okta (for example) can still be used as federated auth, I don't see why is should be exclusive ya know?

File: useragents.txt (hunk 4 above)
Comment (knavesec) [reply useragents.txt:1]: Alright, I'm down. Can we sort -u these and get the count up to 15 just for some additional values?

File: utils/mfasweep.py (hunk 6 above)
Comment (knavesec) [reply utils/mfasweep.py:34]: dafuqqq that's wild hah

File: useragents.txt (hunk 4 above)
Comment (ZephrFish) [reply useragents.txt]: Done!

**Comment by er4z0r** [comment]
Any chance of this making it into master?

**Comment by er4z0r** [comment]
Any chance of this making it into master?
